runtime/shared_common/CacheMap.cpp

```diff
@@ -7453,3 +7683,249 @@ SH_CacheMap::isAddressInReleasedMetaDataBounds(J9VMThread* currentThread, UDATA
 	return rc;
 }
 
+/**
+ *  Get the unique ID of a pre-requisite cache at a lower layer
+ *
+ *	@param [in] currentThread The current JVM thread
+ *	@pamra [in] the cache directory
+ *	@param [in] ccToUse The current cache that depends on lower layer cache.
+ *	@param [in] startupForStats If the cache is started up for cache statistics
+ * 	@param [out] the Unique ID of a re-requiste cache.
+ * 	@param [out] the length of the Unique ID string.
+ *
+ *	@return 1 This cache depends on a lower layer cache. The unique ID of the re-requisite cache is found in or stored to this cache as metadata.
+ *			0 This cache does not depend on a low layer cache.
+ *			CM_CACHE_CORRUPT if cache is corrupted.
+ *			CM_READ_CACHE_FAILED if failed to get the existing pre-requisite cache ID from metadata.
+ *			CM_CACHE_STORE_PREREQ_ID_FAILED if failed to store the pre-requisite cache ID
+ *
+ *
+ */
+IDATA
+SH_CacheMap::getPrereqCache(J9VMThread* currentThread, const char* cacheDir, SH_CompositeCacheImpl* ccToUse, bool startupForStats, const char** prereqCacheID, UDATA* idLen)
+{
+	ShcItem* it = NULL;
+	IDATA result = 0;
+	SH_Manager* manager = NULL;
+	bool isReadOnly = ccToUse->isRunningReadOnly();
+	IDATA rc = 0;
+	PORT_ACCESS_FROM_PORT(_portlib);
+
+	Trc_SHR_Assert_True(ccToUse->hasWriteMutex(currentThread));
+
+	if ((UnitTest::CORRUPT_CACHE_TEST == UnitTest::unitTest)) {
+		return 0;
+	}
+
+	/* If there is a pre-requite cache, the re-requites cache ID the first metadate item */
+	it = (ShcItem*)ccToUse->nextEntry(currentThread, NULL);
+	if (it) {
+		UDATA itemType = ITEMTYPE(it);
+		if ((itemType <= TYPE_UNINITIALIZED) || (itemType > MAX_DATA_TYPES)) {
+			CACHEMAP_TRACE1(J9SHR_VERBOSEFLAG_ENABLE_VERBOSE_DEFAULT, J9NLS_ERROR, J9NLS_SHRC_CM_READ_CORRUPT_DATA, it);
+			if (startupForStats == false && isReadOnly) {
+				ccToUse->setCorruptCache(currentThread, ITEM_TYPE_CORRUPT, (UDATA)it);
+			}
+			reportCorruptCache(currentThread, ccToUse);
+			/* TODO @hangshao Add NLS */
```

Comment:
We shouldn't have TODO's. Either fix or open an Issue.

Reply:
NLS is already there, so removed this TODO